pkg/operator/controller/machineconfig/controller.go

```diff
@@ -0,0 +1,360 @@
+limitations under the License.
+*/
+
+package machineconfigcontroller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/go-logr/logr"
+	mcv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+
+	v1alpha1 "github.com/openshift/node-observability-operator/api/v1alpha1"
+)
+
+// MachineconfigReconciler reconciles a Machineconfig object
+type MachineconfigReconciler struct {
+	client.Client
+
+	Scheme         *runtime.Scheme
+	Log            logr.Logger
+	CtrlConfig     *v1alpha1.Machineconfig
+	EventRecorder  record.EventRecorder
+	PrevSyncChange map[string]PrevSyncData
+}
+
+type PrevSyncData struct {
+	action string
+	config interface{}
+}
+
+const (
+	// MCAPIVersion is the machine config API version
+	MCAPIVersion = "machineconfiguration.openshift.io/v1"
+
+	// MCKind is the machine config resource kind
+	MCKind = "MachineConfig"
+
+	// MCPoolKind is the machine config pool resource king
+	MCPoolKind = "MachineConfigPool"
+
+	// ProfilingMCPName is the name of MCP created for
+	// CRI-O, Kubelet... machine configs by this controller
+	ProfilingMCPName = "profiling"
+)
+
+var (
+	// ProfilingMCSelectorLabels is for storing the labels to
+	// match with profiling MCP
+	ProfilingMCSelectorLabels = map[string]string{
+		"machineconfiguration.openshift.io/role": ProfilingMCPName,
+	}
+)
+
+//+kubebuilder:rbac:groups=nodeobservability.olm.openshift.io,resources=machineconfigs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=nodeobservability.olm.openshift.io,resources=machineconfigs/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=nodeobservability.olm.openshift.io,resources=machineconfigs/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Machineconfig object Gagainst the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+func (r *MachineconfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.Log.Info("Reconciling MachineConfig of Nodeobservability operator")
+
+	// Fetch the nodeobservability.olm.openshift.io/machineconfig CR
+	r.CtrlConfig = &v1alpha1.Machineconfig{}
+	err := r.Get(ctx, req.NamespacedName, r.CtrlConfig)
+	if err != nil {
+		if errors.IsNotFound(err) {
```

Comment:
So, how do we disable profiling? I guess by setting the machine config fields to `false`. But if we want to return the cluster into the original state, I assume we want to delete the machine config generally. In that case we probably want to remove the machine config pool as well, right?

Reply:
This is a good point @saschagrunert - removing machine config and pool once fields are set to false and the cluster returns to original state.

Reply:
Yeah. To enable or disable profiling nodeobservability.olm.openshift.io/v1alpha1/Machineconfig needs to be updated.
Cleaning up of MCP incorporated in [6e77c7b](https://github.com/openshift/node-observability-operator/pull/16/commits/6e77c7b3116158c01e2e14500d27c633e7997885)